src/nvim/misc1.c

```diff
@@ -1067,3 +1067,32 @@ void add_time(char_u *buf, size_t buflen, time_t tt)
                  seconds);
   }
 }
+
+/// Fires a ModeChanged autocmd.
+void trigger_modechanged(void)
+{
+  if (!has_event(EVENT_MODECHANGED)) {
+    return;
+  }
+
+  dict_T *v_event = get_vim_var_dict(VV_EVENT);
+
+  char *mode = get_mode();
```

Comment:
In your original vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?

Also, `get_mode` is in `state.c`, maybe you should put this function in `state.c` as well.

Reply:
> In your original Vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?
> 
> Also, `get_mode` is in `state.c`, maybe you should put this function in `state.c` as well.

I guess your original Vim patch was wrong, as you allocated `MODE_MAX_LENGTH` bytes for `last_mode`.

Reply:
> In your original Vim patch you passed `VAR_UNKNOWN` to `f_mode` so it only returns one character, but here you get the full mode string. Is one of them wrong?

I guess you passed wrong arguments to `f_mode` in your original Vim patch, as you allocated `MODE_MAX_LENGTH` bytes for `last_mode`.